docs/design/flux-sources.md

```diff
@@ -0,0 +1,175 @@
+
+## Motivation
+
+[Flux](https://fluxcd.io/) is a set of Kubernetes controllers that implement GitOps automation. The
+core controllers fetch configuration from a git repository or other such source, and apply it into
+the cluster. Auxiliary controllers handle incoming webhooks (e.g., from GitHub), and send outbound
+notifications (e.g., to Slack).
+
+Broadly speaking, integration with Flux would bring in features that would be expensive to replicate
+in the Pulumi operator. And since the Flux controllers can be used individually, it's possible to
+integrate with selected parts. This design targets the source controller, which handles fetching
+configurations from git repositories, S3 buckets, and OCI (container image) registries.
+
+The benefit of using this integration will be:
+
+ - it works around some limitations of the go-git-based git fetching built into the operator (via
+   automation API) -- specifically, it cannot work with Azure DevOps and some other platforms
+ - it becomes possible to use Flux's notification-controller and webhooks to make the operator more
+   responsive to new commits
+ - users get access to useful features, e.g., you can follow tags according to a semver range, as
+   well as branches; exclude files from the source; verify source revisions using signatures; and so
+   on
+ - users get access to sources other than git repositories: S3 buckets and OCI images
+ - sources and stacks can be specified by different teams, since they are different objects and can
+   have different RBAC rules in effect.
+
+The main disadvantage to using the integration is that you have to run the Flux source controller,
+which is an additional installation step, and another moving part. It is at least simple to install
+and self-contained; that is, there are no dependencies you also need to install and run.
+
+## Design
+
+How this should work mechanically is pretty obvious: you refer to a source in your Stack object,
+rather than supplying git particulars, and the controller fetches from the referenced source object
+rather than from git directly.
+
+### Stack spec
+
+The slightly tricky bit is how to fit this into the Stack API. The git particulars to which the
+sourceRef is an alternative are fields directly under `.spec`, so it is not possible cleanly to use
+[the "union"
+convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#unions),
+in which the fields would be nested in another struct, looking a bit like a tagged union.
+
+```YAML
+# how it _should_ look, but won't ...
+spec:
+  git:                 # )
+    projectRepo: ...   # )
+    gitAuth:           # ) this stanza;
+      # ...            # ) OR,
+
+  source:              # ) this stanza
+    sourceRef:         # )
+      name: foo        # )
+
+  repoDir: ./project # then this field
+```
+
+They can't be moved without manufacturing another version of the API -- probably `v2alpha1` -- which
```

Comment:
Unless a new version is prohibitively expensive, I'd prefer starting with a clean API (i.e., new version) rather than pushing additional complexity to users to handle the retrofit. 

Flux (and maybe inline YAML) significantly expand the scope beyond the original specification, so it would be good to make that spec easy to use and understand. Since these are new features rather than changes to existing features, users who don't need the new features could keep using the existing `v1` spec.

What do you think?

Reply:
All things being equal I would much rather start a new API version -- there are a lot of things that could be tidied up, and more room could be made for new use cases, like this one and inline programs.

However: it needs a lot of planning, and just making a new API doesn't let us throw away the untidy bits -- in fact, it would be much _less_ tidy until the previous API version can be deprecated, because to remain backward compatible it would need all the untidy fields alongside the new fields. This is going to be true whenever we do it, but it does weigh more heavily on the prohibitive expense scales.

I do feel queasy about breaking the usual Kubernetes API convention of not introducing alpha features in a stable version. I think in this case, the usership (and potential usership) may be more interested in the new capabilities than in the convention, so long as they are backward compatible. But it's a marginal call.

Reply:
Makes sense, thanks. One last thought -- have you considered creating a new CRD instead of a new version of the existing Stack CRD? That probably has its own set of problems, but wanted to throw that option out for completeness.

Reply:
> have you considered creating a new CRD instead of a new version of the existing Stack CRD?

I have! Great minds think alike :-) In fact, both that and using a new API group, e.g., `automation.pulumi.com`. The upside of these is that we get to drop fields entirely. The downside is that it puts the work of migration onto each user, who will have to do so manually. A mitigation is that it would be easy to write a tool to migrate individual YAMLs, since the new objects would be backward compatible.